### PR TITLE
config: add security section to schema

### PR DESCRIPTION
**Pull request by Lord-KA**
The following box.cfg options were described in instance_config with defaults similar to ones in box.cfg:
* auth_type
* auth_delay
* disable_guest
* password_lifetime_days
* password_min_length
* password_enforce_uppercase
* password_enforce_lowercase
* password_enforce_digits
* password_enforce_specialchars
* password_history_length

Part of #8861

### CODE REVIEW

File: src/box/lua/config/instance_config.lua [hunk 1]
```diff
@@ -1118,6 +1118,60 @@ return schema.new('instance_config', schema.record({
             end
         end,
     }),
+    security = schema.record({
+        auth_type = schema.enum({
+            'chap-sha1',
+            'pap-sha256',
```
Comment (Totktonada) [inline src/box/lua/config/instance_config.lua:1124]: `pap-sha256` is supported only on Tarantool Enterprise Edition and we should reflect it in the validation.

File: src/box/lua/config/instance_config.lua [hunk 2]
```diff
@@ -1118,6 +1118,60 @@ return schema.new('instance_config', schema.record({
...
+        }, {
+            box_cfg = 'auth_type',
+            default = 'chap-sha1',
+        }),
+        auth_delay = schema.scalar({
+            type = 'number',
+            default = 0,
+            box_cfg = 'auth_delay',
+        }),
```
Comment (Totktonada) [inline src/box/lua/config/instance_config.lua:1133]: This option is supported only on Tarantool Enterprise Edition. We should reflect this fact somehow. Please, check the `enterprise_edition()` fucnction, which should handle it appropriately.

The same for other options of the section.

(I guess it should resolve the test failes in CI.)

File: src/box/lua/config/instance_config.lua (hunk 1 above)
Comment (Lord-KA) [reply src/box/lua/config/instance_config.lua:1124]: Done.

File: src/box/lua/config/instance_config.lua [hunk 3]
```diff
@@ -1118,6 +1118,67 @@ return schema.new('instance_config', schema.record({
...
+            type = 'number',
+            default = 0,
+            box_cfg = 'auth_delay',
+        })),
+        disable_guest = enterprise_edition(schema.scalar({
+            type = 'boolean',
+            default = false,
+            box_cfg = 'disable_guest',
+        })),
+        password_lifetime_days = enterprise_edition(schema.scalar({
```
Comment (ImeevMA) [inline src/box/lua/config/instance_config.lua:1146]: There are quite a few "password" options. What do you think about turning them into `security.password. ...` options? @Lord-KA @Totktonada

File: src/box/lua/config/instance_config.lua [hunk 4]
```diff
@@ -1118,6 +1118,67 @@ return schema.new('instance_config', schema.record({
...
+            'chap-sha1',
+            'pap-sha256',
+        }, {
+            box_cfg = 'auth_type',
+            default = 'chap-sha1',
+            validate = function(auth_type, w)
+                if auth_type ~= 'chap-sha1' and
+                        tarantool.package ~= 'Tarantool Enterprise' then
+                    w.error('chap-sha1 is the only auth_type available in' ..
+                            ' Tarantool CE')
```
Comment (ImeevMA) [inline src/box/lua/config/instance_config.lua:1132]: I think it is better to write something like `The PAP-SHA256 authentication method is only available in Tarantool Enterprise Edition`. This way it will be similar to our usual `This configuration parameter is available only in Tarantool Enterprise Edition`.

**Review by ImeevMA — APPROVED** [review]
Hi! Thank you for the patch! I have a couple of comments, but I don't think they are blockers.

File: src/box/lua/config/instance_config.lua (hunk 3 above)
Comment (Lord-KA) [reply src/box/lua/config/instance_config.lua:1146]: I think it is best to have structure as similar with box.cfg as possible. "security" itself is a made up name, I don't see the necessity for another one.

Comment (ImeevMA) [reply src/box/lua/config/instance_config.lua:1146]: Well, `security.password.lifetime_days` is not too different from `security.password_lifetime_days` I guess. But it looks more structured.

File: src/box/lua/config/instance_config.lua (hunk 4 above)
Comment (ImeevMA) [reply src/box/lua/config/instance_config.lua:1132]: I think it's better to describe what a user would get in EE than to say what we have in CE.

File: src/box/lua/config/instance_config.lua (hunk 3 above)
Comment (Lord-KA) [reply src/box/lua/config/instance_config.lua:1146]: Yea, but `password_lifetime_days` is exactly box.cfg option, while `lifetime_days` isn't.

File: src/box/lua/config/instance_config.lua (hunk 4 above)
Comment (Totktonada) [reply src/box/lua/config/instance_config.lua:1132]: 1. CE/EE abbreviations shouldn't be used in error messages, because it is not obvious for a user who isn't get in touch with the project hard enough to know our internal teminology. Let's use explicit 'Tarantool Community Edition' and 'Tarantool Enterprise Edition'.
2. It would be good to mention the problematic value in the error message.
3. I would prefer to show configuration values as is in quotes (use `%q`) and use human-readable option description to point the option ('authentication method' instead of 'auth_type').
4. The new authentication method may be added into CE, EE, both or moved from EE to CE. We anyway should reflect it in the configuration schema.

To sum up, I mostly support Mergen's proposal.

File: src/box/lua/config/instance_config.lua [hunk 5]
```diff
@@ -1118,6 +1118,67 @@ return schema.new('instance_config', schema.record({
...
+            default = false,
+            box_cfg = 'password_enforce_digits',
+        })),
+        password_enforce_specialchars = enterprise_edition(schema.scalar({
+            type = 'boolean',
+            default = false,
+            box_cfg = 'password_enforce_specialchars',
+        })),
+        password_history_length = enterprise_edition(schema.scalar({
+            type = 'number',
```
Comment (Totktonada) [inline src/box/lua/config/instance_config.lua:1177]: Can we use `type = 'integer'` here?

File: test/config-luatest/instance_config_schema_test.lua [hunk 6]
```diff
@@ -967,3 +967,25 @@ g.test_app = function()
     local res = instance_config:apply_default({}).app
     t.assert_equals(res, nil)
 end
+
+g.test_security = function()
```
Comment (Totktonada) [inline test/config-luatest/instance_config_schema_test.lua:971]: I'm looking into PR #8895 as well and there are two good ideas in it:

* Add a test that starts a real server and verifies `box.cfg.<...>` values (see config_test.lua).
* Add enterprise/community `:validate()` and `:apply_default()` test cases both. The community one would have different defaults and would receive an error on attempt to set some options/values.

File: src/box/lua/config/instance_config.lua [hunk 7]
```diff
@@ -1118,6 +1118,67 @@ return schema.new('instance_config', schema.record({
             end
         end,
     }),
+    security = schema.record({
+        auth_type = schema.enum({
+            'chap-sha1',
+            'pap-sha256',
+        }, {
+            box_cfg = 'auth_type',
+            default = 'chap-sha1',
```
Comment (Totktonada) [inline src/box/lua/config/instance_config.lua:1127]: AFAIU, you said that the defaults are different on CE and EE and we agreed to set `default = box.NULL` in the schema with a comment about the meaning of such a value.

File: src/box/lua/config/instance_config.lua [hunk 8]
```diff
@@ -1118,6 +1118,67 @@ return schema.new('instance_config', schema.record({
             end
```
Comment (Totktonada) [inline src/box/lua/config/instance_config.lua:1118]: > NO_DOC=in the subsequent patches in scope of #8861
> NO_CHANGELOG=in the subsequent patches in scope of #8861

Please, add a changelog entry.

Please, add more real reason to `NO_DOC`. Say,

> NO_DOC=https://github.com/tarantool/doc/issues/3544 links the most actual schema, no need to update the issue.

File: src/box/lua/config/instance_config.lua (hunk 4 above)
Comment (Lord-KA) [reply src/box/lua/config/instance_config.lua:1132]: @Totktonada, I mostly agree with you, see comments below:
1) Fixed.
2) Done
3) Agreed.
4) The validation function shouldn't really have such info other than the list of options that are valid. Now there is only one right option, so there is no need for complexity. I don't mind it being added in the future with other `auth_type`-s.

```lua
              validate = function(auth_type, w)
                  if auth_type ~= 'chap-sha1' and
                          tarantool.package ~= 'Tarantool Enterprise' then
                          w.error("'chap-sha1' is the only auth_type available in" .. 
                                  " Tarantool Community Edition (%q requested)",
                                  auth_type)                                                                  
                  end                                                                                                                                                              
              end                                                                                                                                                                                                                                                                                                                                
```

File: src/box/lua/config/instance_config.lua (hunk 5 above)
Comment (Lord-KA) [reply src/box/lua/config/instance_config.lua:1177]: Fixed. Same for `password_lifetime_days`, according to [doc](https://www.tarantool.io/en/enterprise_doc/security/#confval-password_lifetime_days).

Comment (Lord-KA) [reply src/box/lua/config/instance_config.lua:1177]: (but `auth_delay` may be non-integer)

File: src/box/lua/config/instance_config.lua (hunk 8 above)
[the remark this answers is not in the document]
Comment (Lord-KA) [reply src/box/lua/config/instance_config.lua:1118]: Done.

File: src/box/lua/config/instance_config.lua (hunk 7 above)
Comment (Lord-KA) [reply src/box/lua/config/instance_config.lua:1127]: I think, there was a misunderstanding, the default in both CE and EE [is `chap-sha1`](https://www.tarantool.io/en/enterprise_doc/security/#authentication-protocol), my problem was that I didn't know how to implement validation and you helped me with that.

File: src/box/lua/config/instance_config.lua (hunk 8 above)
Comment (Lord-KA) [reply src/box/lua/config/instance_config.lua:1118]: > Please, add a changelog entry.

Ok. But I don't understand how it is helpful. I would say one changelog entry with "added all box.cfg options to config" would be more than enough.

File: test/config-luatest/instance_config_schema_test.lua (hunk 6 above)
Comment (Lord-KA) [reply test/config-luatest/instance_config_schema_test.lua:971]: > Add a test that starts a real server and verifies box.cfg.<...> values (see config_test.lua).

If you intend to request such test on every section, it needs some gentrification. Copy-pasting minimal config every time is meaningless.

Comment (Lord-KA) [reply test/config-luatest/instance_config_schema_test.lua:971]: > Add enterprise/community :validate() and :apply_default() test cases both. The community one would have different defaults and would receive an error on attempt to set some options/values.

We already do test the defaults in `cluster_config_schema_test.lua`, both EE and CE. Added test to check validation error in CE.

File: changelogs/unreleased/gh-8861-security-options.md [hunk 9]
```diff
@@ -0,0 +1,3 @@
+## feature/config
+
+* All security hardening box.cfg options are now supported by config (gh-8861).
```
Comment (xuniq) [inline changelogs/unreleased/gh-8861-security-options.md:3]:
```suggestion
* All security hardening `box.cfg{}` options are now supported by config (gh-8861).
```

**Review by xuniq — APPROVED** [review]

File: changelogs/unreleased/gh-8861-security-options.md (hunk 9 above)
Comment (Lord-KA) [reply changelogs/unreleased/gh-8861-security-options.md:3]: Fixed.

File: \ [hunk 10]
```diff
@@ -0,0 +1,4 @@
+## feature/config
```
Comment (Totktonada) [inline \:1]: Please, look at the file name.

File: src/box/lua/config/instance_config.lua (hunk 4 above)
Comment (Totktonada) [reply src/box/lua/config/instance_config.lua]: It is good enough for my taste, but I would use 'authentication method' instead of 'auth_type' (see point 3 above).

File: src/box/lua/config/instance_config.lua (hunk 5 above)
Comment (Totktonada) [reply src/box/lua/config/instance_config.lua]: Regarding `password_lifetime_days`: I looked over the code and it seems that nothing prevents us from setting a floating point value. However, it has no much sense. I'm OK with `integer` here.

File: src/box/lua/config/instance_config.lua (hunk 8 above)
Comment (Totktonada) [reply src/box/lua/config/instance_config.lua]: Changelogs with certain changes may be useful, when you're in a trouble and want to do some guesses about a reason of the trouble based on some reading, without real bisect effort. Sometimes it saves some time. It is a good thing to have, especially for a library or a developer platform.

Not a critical thing, but a nice thing and we all agreed to do that in tarantool.

File: test/config-luatest/config_test.lua [hunk 11]
```diff
@@ -391,3 +391,59 @@ g.test_feedback_options = function()
         t.assert_equals(box.cfg.feedback_metrics_limit, 1000000)
     end)
 end
+
+g.test_security_options = function()
+    t.tarantool.skip_if_not_enterprise()
+    local dir = treegen.prepare_directory(g, {}, {})
+    -- guest user is required for luatest.server helper to function properly,
+    -- so it is enabled (`disable_guest: false`) unlike other options.
```
Comment (Totktonada) [inline test/config-luatest/config_test.lua:399]: We can grant `super` on another user and pass `net_box_credentials = {user = <...>, password = <...>}` to `server:new()` if you like. There is nothing that would actually require `guest` user.

File: test/config-luatest/instance_config_schema_test.lua (hunk 6 above)
Comment (Totktonada) [reply test/config-luatest/instance_config_schema_test.lua]: It is my personal feeling that without a kind of end-to-end scenario in the automatic testing we can broke some options without noticing it. This is a weak requirement, but it is a ground level I would require from these patches.

I looked over the patch and now the testing looks quite good. Thanks!

File: test/config-luatest/instance_config_schema_test.lua [hunk 12]
```diff
@@ -1210,3 +1198,54 @@ g.test_feedback_disabled = function()
                 'without feedback reports sending support'
     t.assert_equals(err, exp)
 end
+
+g.test_security_enterprise = function()
+    t.tarantool.skip_if_not_enterprise()
+
```
Comment (Totktonada) [inline test/config-luatest/instance_config_schema_test.lua:1204]: Nit: I would add one success scenario for CE (just `security.auth_type = 'chap-sha1'`).

**Review by Totktonada — APPROVED** [review]
OK in general, but please consider several small remarks.

File: test/config-luatest/config_test.lua (hunk 11 above)
Comment (Lord-KA) [reply test/config-luatest/config_test.lua:399]: It isn't necessary, I wrote the comment to explain why I used default value here, unlike other options.

Comment (Totktonada) [reply test/config-luatest/config_test.lua:399]: OK, it is up to you. I'm going to resolve the discussion.

File: \ (hunk 10 above)
Comment (Lord-KA) [reply \:1]: It happens with vim `:w` missclick, removed.

File: test/config-luatest/instance_config_schema_test.lua (hunk 12 above)
Comment (Lord-KA) [reply test/config-luatest/instance_config_schema_test.lua:1204]: Done.
